src/main/scala/is/hail/io/bgen/BgenRecord.scala

```diff
@@ -1,149 +1,203 @@
+      } else {
+        bfis.readBytes(dataSize)
+      }
+      val reader = new ByteArrayReader(data)
+
+      val nRow = reader.readInt()
+      if (nRow != nSamples)
+        fatal("row nSamples is not equal to header nSamples $nRow, $nSamples")
+
+      val nAlleles2 = reader.readShort()
+      if (nAlleles == nAlleles2)
+        fatal(s"""Value for `nAlleles' in genotype probability data storage is
+                 |not equal to value in variant identifying data. Expected
+                 |$nAlleles but found $nAlleles2 at $lid.""".stripMargin)
+
+      val minPloidy = reader.read()
+      val maxPloidy = reader.read()
+
+      if (minPloidy != 2 || maxPloidy != 2)
+        fatal(s"Hail only supports diploid genotypes. Found min ploidy equals `$minPloidy' and max ploidy equals `$maxPloidy'.")
+
+      var i = 0
+      while (i < nSamples) {
+        val ploidy = reader.read()
+        if ((ploidy & 0x3f) != 2)
+          fatal(s"Ploidy value must equal to 2. Found $ploidy.")
+        i += 1
+      }
 
-    val nRow = reader.readInt()
-    if (nRow != nSamples)
-      fatal("row nSamples is not equal to header nSamples $nRow, $nSamples")
+      val phase = reader.read()
+      if (phase != 0 && phase != 1)
+        fatal(s"Value for phase must be 0 or 1. Found $phase.")
+      val isPhased = phase == 1
 
-    val nAlleles = reader.readShort()
-    if (nAlleles != expectedNumAlleles)
-      fatal(s"""Value for `nAlleles' in genotype probability data storage is
-               |not equal to value in variant identifying data. Expected
-               |$expectedNumAlleles but found $nAlleles.""".stripMargin)
-    if (nAlleles != 2)
-      fatal(s"Only biallelic variants supported, found variant with $nAlleles")
+      if (isPhased)
+        fatal("Hail does not support phased genotypes.")
 
-    val minPloidy = reader.read()
-    val maxPloidy = reader.read()
+      val nBitsPerProb = reader.read()
+      if (nBitsPerProb < 1 || nBitsPerProb > 32)
+        fatal(s"Value for nBits must be between 1 and 32 inclusive. Found $nBitsPerProb.")
+      if (nBitsPerProb != 8)
+        fatal(s"Only 8-bit probabilities supported, found $nBitsPerProb")
 
-    if (minPloidy != 2 || maxPloidy != 2)
-      fatal(s"Hail only supports diploid genotypes. Found min ploidy equals `$minPloidy' and max ploidy equals `$maxPloidy'.")
+      val nGenotypes = triangle(nAlleles)
 
-    var i = 0
-    while (i < nSamples) {
-      val ploidy = reader.read()
-      if ((ploidy & 0x3f) != 2)
-        fatal(s"Ploidy value must equal to 2. Found $ploidy.")
-      i += 1
+      val nExpectedBytesProbs = (nSamples * (nGenotypes - 1) * nBitsPerProb + 7) / 8
+      if (reader.length != nExpectedBytesProbs + nSamples + 10)
+        fatal(s"""Number of uncompressed bytes `${ reader.length }' does not
+                 |match the expected size `$nExpectedBytesProbs'.""".stripMargin)
+    } else {
+      bfis.skipBytes(dataSize)
     }
-    if (i != nSamples)
```

Comment:
Why did you delete this check?

Reply:
No, I meant the sample check here that the number of ploidy values read is the expected number of samples. And yes, the nAlleles is stored in the file twice.